File: Sliding Window/340. Longest Substring with At Most K Distinct Characters.py
class Solution:
    def lengthOfLongestSubstringKDistinct(self, s: str, k: int) -> int:
        chars = collections.defaultdict(int)
        l = 0
        n = len(s)
        res = 0
        for r in range(n):
            chars[s[r]] += 1
            while len(chars) > k:
                chars[s[l]] -= 1
                if chars[s[l]] == 0:
                    del chars[s[l]]
                l += 1
            res = max(res, r - l + 1)
        return res

File: Sliding Window/340. Longest Substring with At Most K Distinct Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstringKDistinct(string s, int k) {
        unordered_map<char, int>mp;
        int c = 0;
        int n = s.length();
        int l = 0;
        int res = 0;
        for (int i = 0; i < n; i++) {
            char cs= s[i];
            mp[cs]++;
            if (mp[cs] == 1) c++;
            while (c > k) {
                mp[s[l]]--;
                if (!mp[s[l]]) {
                    c--;
                }
                l++;
            }
            res = max(res, i - l + 1);
        }
        return res;

    }
};